### ISSUE TITLE
Migrate bots to GitHub Actions

### ISSUE THREAD

**Issue by timvandermeij** [mozilla/pdf.js#11851]
PDF.js has two dedicated bots for testing. However, this is all custom and maintained manually. This is already a problem now since the bots run quite old OS versions that lack a (modern enough) Python 3 version, which blocks the `ttx` upgrade for example (see #11802).

We could unblock this work by upgrading the OS versions, but for the longer term it would be better for the project to move towards a managed solution such as GitHub Actions. The latter integrates well with GitHub and our existing jobs, and reading quickly through the docs might allow us to create a `botio`-like flow where whitelisted users can comment on PRs to trigger actions (builds).

In terms of our builds, I mainly see a challenge for the browser tests since they are stateful, so we need to be able to fetch/store the reference images somewhere. I'm thinking this could be done in an S3 bucket, and let GitHub manage the credentials for that.

Let's investigate if these solutions could work for us and how much logic we have to maintain ourselves. The objective is to use existing solutions as much as possible so we don't have to build/maintain it ourselves, as well as improving reliability.

**Comment by timvandermeij** [comment]
GitHub Actions might actually be available for us now. I think it wasn't before, but now https://github.com/mozilla/pdf.js/actions/new is available, even though the actions tab in the tab bar above is not available. I think the project owner can enable that.

**Comment by timvandermeij** [comment]
The switch to GitHub Actions is now made in #12760 because of #12743.

**Comment by brendandahl** [comment]
For the long term health of the project I think it would be good to migrate all testing to github actions. As Tim mentioned we can store refs in S3 or alternatively Marco mentioned gh action artifacts can be stored for 90 days so we could set up a cron job to maintain them.

**Comment by timvandermeij** [comment]
I think using GitHub artifacts might also be a good and perhaps even easier solution since everything then stays within the GitHub ecosystem, so we don't have to manage another AWS account for S3. The cronjob should be doable.

**Comment by timvandermeij** [comment]
Time permitting I'll try to set up a basic workflow to run the bot commands using GitHub Actions, without artifact retention for now, so we can see how/if it works and go from there.

**Comment by timvandermeij** [comment]
I have been working on migrating the preview task from the bot to GitHub Actions in https://github.com/timvandermeij/pdf.js/commit/fe60c41b69549c48a658b898afb49d00c57db39f and while mostly working, I did run into the limitation that build artifact files can't seem to be accessed individually, at least through the UI, since the are all combined into one ZIP file; see the example run at https://github.com/timvandermeij/pdf.js/actions/runs/1085127345. There are actions to post a comment on a PR with a link to the artifact, but if we can't get a direct link that opens the viewer in the browser that'll be inconvenient. Something similar will be needed for the reftest analyzer for migrating the reference testing task.

This limitation is already reported inhttps://github.com/actions/upload-artifact/issues/14, so given that GitHub doesn't really "host" the artifacts separately, we might still need to use an S3 bucket to make the preview/reftest analyzer available in the browser from a direct link, similar to how Botio does it. We can then also store the reference images there, avoiding the need for a cronjob.

**Comment by timvandermeij** [comment]
I now have a working implementation of Botio's lint command fully in a GitHub Actions workflow in commit https://github.com/timvandermeij/pdf.js/commit/d37aa7d4a09ed808a15950f57cb9ee18ac6c80de. It works the same as Botio: if a specific comment is places on a pull request by an authorized user, it will trigger the command and put notification comments on the pull request. https://github.com/timvandermeij/pdf.js/pull/5#issuecomment-894801711 is an example of how that looks (first a commit that failed linting, then one that made it pass). I think it's pretty cool how this works and it shows how nicely it integrates with GitHub and how it matches the current Botio behavior.

I chose the lint command since it's easiest to set up the issue comment infrastructure with since we don't have to bother with running browsers. However, while Botio has the linting command, I'm not sure if we should also introduce that in the new setup given that linting by default already runs in the CI workflow for every push, so I don't think there is ever a point where we want to run linting on-demand since it already happens (and if it fails, we can simply retrigger the CI build). Therefore, once I get the browsers working, we can start introducing workflows for all Botio tasks that don't require assets since that's pending on the S3 point above (basically everything except for the browser tests).

**Comment by Snuffleupagus** [comment]
One thing that has always annoyed me a little bit with the `botio` setup is the need to have *separate* bot-specific tasks for all the tests we want to run (look at the gulp-tasks whose names start with "bot...").

In the new "GitHub Actions" setup, would it be possible to use command line flags instead together with the "normal" gulp-tasks (to reduce duplication)?
For example, instead of having to run `gulp bottest`, could we run `gulp test --bot` and have `createTestSource` directly check for that flag?

**Comment by timvandermeij** [comment]
I don't see why not. In fact, I'm not really sure why that's even needed in the current Botio setup given that I don't see anything stopping us from using `gulp test --bot` at https://github.com/mozilla/botio-files-pdfjs/blob/master/on_cmd_test.js#L42? Botio or GitHub Actions will simply run the command we provide it, so if the Gulpfile can handle that argument we're good.

**Comment by brendandahl** [comment]
> @brendandahl Would it be possible for you to set up an S3 bucket and configure the required secrets for accessing it in GitHub's secret manager? Once that's done, we can test uploading our artifacts and reference images there. See https://dev.to/johnkevinlosito/deploy-static-website-to-s3-using-github-actions-4a0e for how to do this.

Yeah, we should be able to do that. Let me look into getting it set up.

**Comment by brendandahl** [comment]
@timvandermeij I've added an s3 bucket. I've also added the following repository secrets:

AWS_ACCESS_KEY_ID
AWS_REGION
AWS_S3_BUCKET
AWS_SECRET_ACCESS_KEY

**Comment by timvandermeij** [comment]
Nice; thank you! It will require some more tweaking of my initial GitHub Actions workflows, but I think we now have all the pieces of the puzzle to implement all features that our current bots have.

**Comment by marco-c** [comment]
There are things like https://github.com/marketplace/actions/debugging-with-ssh that would allow us to access the GH Actions machines in an interactive fashion through ssh.

**Comment by marco-c** [comment]
@timvandermeij do you think you will have time to continue this?

**Comment by timvandermeij** [comment]
Sadly I won't have much time on the short term. If you want to have it done faster, feel free to take the proof-of-concept work in https://github.com/mozilla/pdf.js/issues/11851#issuecomment-894803689 and go from there. It mainly needed some more thought given that GitHub did not have a way of inheriting parts of jobs, so for example the list of allowed users would have to be duplicated in every job file, which seems suboptimal. Moreover, we need to try if storing the reference images in S3 works for the browser tests.

**Comment by Snuffleupagus** [comment]
Re-reading the discussion above it seems that most of "problems" are around testing in particular, so could we perhaps make easier progress here by ignoring that part for now?
Basically, could we start by only moving tasks like building the `gh-pages` branch and the `pdfjs-dist` library from the bots and into GitHub actions instead?

**Comment by timvandermeij** [comment]
The source code for the existing commands of the bots can be found at https://github.com/mozilla/botio-files-pdfjs. The list of commands of the bots is as follows, and all of them should get an equivalent GitHub Actions workflow before we can shut down the bots.

- Non-testing commands:
   - [x] `on_push.js`: publishes a new website on every push to the `master` branch.
      - Fixed in #16364 and https://github.com/mozilla/botio-files-pdfjs/pull/39.
   - [x] `on_release.js`: publishes a new release to the `pdfjs-dist` repository and NPM.
      - Fixed in #18350 for NPM and filed #18357 for the `pdfjs-dist` repository.
- Testing commands:
   - [ ] `on_cmd_browsertest.js`: runs the reference tests.
   - [x] `on_cmd_fonttest.js`: runs the font (TTX) tests.
     - Fixed in #17263 and https://github.com/mozilla/botio-files-pdfjs/pull/42.
   - [x] `on_cmd_integrationtest.js`: runs the integration tests.
     - Fixed in https://github.com/mozilla/pdf.js/pull/20371 and https://github.com/mozilla/botio-files-pdfjs/pull/48.
   - [x] `on_cmd_lint.js`: runs the linting command.
     - Fixed in https://github.com/mozilla/pdf.js/pull/12760 and https://github.com/mozilla/botio-files-pdfjs/pull/41.
   - [ ] `on_cmd_makeref.js`: runs the reference image generation.
     - Reduced in https://github.com/mozilla/botio-files-pdfjs/pull/44.
   - [ ] `on_cmd_preview.js`: builds a preview of the library and viewer for inspection in the browser.
   - [x] `on_cmd_test.js`: runs the unit, font, integration and reference tests.
     - Fixed in https://github.com/mozilla/botio-files-pdfjs/pull/48.
   - [x] `on_cmd_unittest.js`: runs the unit tests.
     - Fixed in https://github.com/mozilla/pdf.js/pull/21121 and https://github.com/mozilla/botio-files-pdfjs/pull/48.
   - [x] `on_cmd_xfatest.js`: runs the unit, integration and reference test, but for the latter only the reference tests that involve XFA (as a shortcut to not have to run all tests for XFA changes).
     - Fixed in https://github.com/mozilla/pdf.js/pull/20882 and https://github.com/mozilla/botio-files-pdfjs/pull/47.

**Comment by Snuffleupagus** [comment]
> * `on_release.js`: publishes a new release to the `pdfjs-dist` repository and NPM.

That seems like the single most important one to me, given all of the problems we've had with recent release. (Also, we really "need" to do a new release given some of the standard-font data stuff being worked on now.)

> * `on_cmd_lint.js`: runs the linting command.

https://github.com/mozilla/pdf.js/issues/11851#issuecomment-894803689 talks about possibly just removing that one, which is something that I'd fully support.

**Comment by marco-c** [comment]
Thinking about this again, it might be easier to store reference images in a GitHub repository. This way we can easily access images from any run of makeref, and we don't need to set up anything in S3 or GCS.

**Comment by timvandermeij** [comment]
> Thinking about this again, it might be easier to store reference images in a GitHub repository. This way we can easily access images from any run of makeref, and we don't need to set up anything in S3 or GCS.

I think that could work, but then we still need a location for storing the cached PDF files, and especially the linked PDFs are difficult because we don't want to download them again for every run but we also can't put them in a Git repository (for the same reason we currently don't and have the concept of linked PDFs in the first place). Currently we cache them on the filesystem on the bots, and we will need some equivalent option for GitHub Actions.

Do you have ideas on this? (It has been a while since I worked on it so I'm not sure if the secrets listed in https://github.com/mozilla/pdf.js/issues/11851#issuecomment-911955059 are still valid, but lately we've been making more progress in moving bot actions to GitHub Actions so this becomes relevant again.)

**Comment by marco-c** [comment]
What if we used a private github repository? It would be essentially the same as a local filesystem.

**Comment by timvandermeij** [comment]
Yes, I think that can definitely work. It could store the PDF files that only GitHub Actions may fetch with special credentials, and maybe setting up GitHub Pages for that repository could give us a way to expose e.g. the reference images viewer and the previews we build using a public URL similar to how Botio currently works.

**Comment by Snuffleupagus** [comment]
It seems that any further progress on this issue is now completely blocked by having *somewhere* to store test PDFs and reference images.

Most likely we'll need to do some "real world" experimentation/testing to see what will/won't work, and converting e.g. `botio unittest` might be a good place to start; however based on the recent discussion we'll need:
 - A *private* GitHub repository setup. (It'd be nice if Tim and myself had access to it, to make things easier.)
 - The private repository configured such that the PDF.js GitHub Actions can access it, as mentioned in https://github.com/mozilla/pdf.js/issues/11851#issuecomment-1817487963, for both reading *and* writing.

@calixteman, @marco-c Can either of you please help out with the above?

With that in place we'd *hopefully* be able to replicate the reading/writing that `botio unittest` currently does, see
 - https://github.com/mozilla/botio-files-pdfjs/blob/a514bbfe2d8a71cda350b6c404028520784ff29d/on_cmd_unittest.js#L10-L15
 - https://github.com/mozilla/botio-files-pdfjs/blob/a514bbfe2d8a71cda350b6c404028520784ff29d/on_cmd_unittest.js#L40-L46

which again hopefully would be enough to migrate the unit-tests to GitHub Actions. Once this works, the other test-suites should be fairly simple to convert as well.

**Comment by marco-c** [comment]
I'm going to create a pdf.js-references repository and give all of us access.

**Comment by marco-c** [comment]
Regarding accessing the repo through GitHub Actions, we can do it through the @moz-tools-bot bot. We can create a secret that stores the API key and push to the repo by doing something like "https://{git_user}:{git_password}@github.com/{repo}".

**Comment by marco-c** [comment]
>     * Testing commands:
>       * [ ]  `on_cmd_browsertest.js`: runs the reference tests.
>       * [ ]  `on_cmd_integrationtest.js`: runs the integration tests.
>       * [ ]  `on_cmd_makeref.js`: runs the reference image generation.
>       * [ ]  `on_cmd_preview.js`: builds a preview of the library and viewer for inspection in the browser.
>       * [ ]  `on_cmd_test.js`: runs the unit, font, integration and reference tests.
>       * [ ]  `on_cmd_unittest.js`: runs the unit tests.

Is there any other test command that would be easy to move to GH Actions?  E.g. maybe the unittest one?

>       * [ ]  `on_cmd_xfatest.js`: runs the unit, integration and reference test, but for the latter only the reference tests that involve XFA (as a shortcut to not have to run all tests for XFA changes).

Given we're not doing a lot of XFA changes anymore, perhaps we could remove this command at this point.

**Comment by marco-c** [comment]
>       * [ ]  `on_cmd_preview.js`: builds a preview of the library and viewer for inspection in the browser.

For this one we could rely on something like https://github.com/marketplace/actions/deploy-pr-preview maybe.

**Comment by marco-c** [comment]
> > ```
> >   * [ ]  `on_cmd_preview.js`: builds a preview of the library and viewer for inspection in the browser.
> > ```
> 
> For this one we could rely on something like https://github.com/marketplace/actions/deploy-pr-preview maybe.

I had another look, this isn't really usable because it only previews from the main repo and not from forks.

@timvandermeij for unittests, would it be just a matter of adding `npx gulp unittest` to https://github.com/mozilla/pdf.js/blob/master/.github/workflows/ci.yml? I see the bot is just running that, with the only difference that it has a cache. Are there lots of PDFs downloaded during unittest or can we live without a cache? If a cache is strictly required, we could use https://github.com/actions/cache (which could actually also be useful for reftests!).

**Comment by timvandermeij** [comment]
The cache solution might work indeed, but a few things come to mind that we need to look into

- The lifetime of the cache is relatively short. Looking at https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy it seems to keep the cache for at most 7 days. We may be able to work around that by creating a scheduled build (if that's possible) to keep the cache "warm" and prevent it from being deleted.
- We need to make sure that this works properly w.r.t. the cache key: the bots guarantee that at most one job runs at a time, which ensures cache integrity if e.g. the first job updates it, but on GitHub Actions we can have builds running in parallel, so we need to be careful that one job does not overwrite the cache of another job. This might be solvable by setting concurrency to 1 (likely possible looking at https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/control-the-concurrency-of-workflows-and-jobs).

I'm going to try to think about this a bit more during the weekend to make sure I'm not missing something here, but so far this does seem like the best way forward (might be easier than using a separate Git repository).

**Comment by marco-c** [comment]
We could use a cache key that contains the hashes of the files in the pdfs directory, this way whenever one file in that directory changes we re-run the script to generate the cache. This way even if there are two jobs running, it won't create problems (they will both download any new PDF file, with a small waste of resources, but the cache will still work).
Then, we could use restore-keys to download the previous version of the cache with the previous already downloaded PDFs.

Similar to https://github.com/actions/cache/blob/main/caching-strategies.md#using-restore-keys-to-download-the-closest-matching-cache.

**Comment by timvandermeij** [comment]
This sounds like it might work. My current focus is on making the integration tests run independent of the order, to reduce the impact of and aid fixing the remaining intermittents we have because those might cause issues to go by unnoticed in PRs, but after that I can give this a try.

**Comment by timvandermeij** [comment]
The current status is that everything is migrated from the bots to GitHub Actions except for the preview, browsertest and makeref targets. The browsertest target is currently being worked on by @calixteman.

**Comment by marco-c** [comment]
"preview" should be simpler than browsertest and makeref, so we are very close here finally!
